core/formats/formats_burst_trie.cpp

```diff
@@ -54,30 +58,286 @@
   // NOOP
 #endif
 
-#include "shared.hpp"
 #include "format_utils.hpp"
 #include "analysis/token_attributes.hpp"
+#include "formats/formats_10_attributes.hpp"
 #include "index/iterators.hpp"
 #include "index/index_meta.hpp"
 #include "index/field_meta.hpp"
 #include "index/file_names.hpp"
 #include "index/index_meta.hpp"
+#include "store/memory_directory.hpp"
+#include "store/store_utils.hpp"
 #include "utils/automaton.hpp"
+#include "utils/buffers.hpp"
+#include "utils/encryption.hpp"
+#include "utils/hash_utils.hpp"
+#include "utils/memory.hpp"
+#include "utils/memory_pool.hpp"
+#include "utils/noncopyable.hpp"
 #include "utils/directory_utils.hpp"
+#include "utils/fstext/fst_string_weight.h"
 #include "utils/fstext/fst_builder.hpp"
+#include "utils/fstext/fst_decl.hpp"
 #include "utils/fstext/fst_matcher.hpp"
+#include "utils/fstext/fst_string_ref_weight.h"
 #include "utils/fstext/fst_table_matcher.hpp"
+#include "utils/fstext/immutable_fst.h"
 #include "utils/timer_utils.hpp"
 #include "utils/bit_utils.hpp"
 #include "utils/bitset.hpp"
 #include "utils/frozen_attributes.hpp"
 #include "utils/string.hpp"
 #include "utils/log.hpp"
 
-namespace {
+namespace  {
+
+using namespace irs;
+
+template<typename Char>
+class volatile_ref : util::noncopyable {
+ public:
+  typedef irs::basic_string_ref<Char> ref_t;
+  typedef std::basic_string<Char> str_t;
+
+  volatile_ref() = default;
+
+  volatile_ref(volatile_ref&& rhs) noexcept
+   : str_(std::move(rhs.str_)),
+     ref_(str_.empty() ? rhs.ref_ : ref_t(str_)) {
+    rhs.ref_ = ref_;
+  }
+
+  volatile_ref& operator=(volatile_ref&& rhs) noexcept {
+    if (this != &rhs) {
+      str_ = std::move(rhs.str_);
+      ref_ = (str_.empty() ? rhs.ref_ : ref_t(str_));
```

Comment:
Is it safe to assume that if str_ is empty it is ref to external ? Especially if non-volatile assign works as implemented (not cleaning str_). It would be more safe to check initially if rhs.ref_ points to rhs.str_.